### ISSUE TITLE
NVDA does not announce focused recipient name it just announces graphic while navigating between recipients added in To/CC/BCC Fields in outlook web

### ISSUE THREAD

**Issue by msftedad** [nvaccess/nvda#16631]
### Steps to reproduce:
1. Open above URL (https://outlook.office.com/)
2. Navigate to New email from Mail tab and activate it.
3. Now navigate to To/CC/BCC field and multiple recipients and try to navigate between the recipients. 
4. Observe screen reader announces only graphic or blank but does not announce the name.

### Actual behavior:

NVDA does not announce focused recipient name it just announces graphic while navigating between recipients added in To/CC/BCC Fields.

### Expected behavior:

NVDA should announce focused recipient name while navigating between recipients added in To/CC/BCC Fields.

### NVDA logs, crash dumps and other attachments:
[NVDA still repro .webm](https://github.com/nvaccess/nvda/assets/93735775/7d03f48e-a023-4beb-bb99-065bfe14e2cd)

#### NVDA installed/portable/running from source: 
Installed

#### NVDA version:
2024.1.0.31547

#### Windows version:
Windows 11 Version 24H2 (Build 26200.5001)

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes. Same behaviour occurs now with NVDA 2023 also

#### If NVDA add-ons are disabled, is your problem still occurring?
NA

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
NA

**Comment by Adriani90** [comment]
Using Firefox and Chromium, In focus mode, NVDA announces "button" and the first letter of the email address of the recipient.

This is because the addresses are wrapped into buttons and the button contains a ZERO WIDTH SPACE character (0x200b) which is rendered as a separate child element. NVDA reports only the first child element, and this is probably the zero width space character and the first letter of the button. I am not sure I understand this correctly.

@jcsteh any ideas how to deal with these zero width space characters? I saw this in several other scenarios and browsers seem to create a separate child in the DOM with these characters preceeding the actual element. This causes NVDA to fail in reporting the element with its label / accessible name.
Maybe NVDA could ignore them in the simplified object hierarchy when presenting objects. Would that fix this issue?

**Comment by jcsteh** [comment]
How are you navigating between recipients?

I'd say the 0 width space is a red herring. I'm not able to test outlook.com right now, but I'm guessing this is to do with the fact that this is an embedded control inside a rich text editor, but the control is being treated as a single character. This happens because left and right arrows in an editor move by character, so NVDA does exactly that: reads by character.

Distilled test case:

`data:text/html,<div contenteditable>a<button>bc</button>d`

Note that if you tab to the text box and then navigate with the left and right arrow keys, NVDA reports a, b and d, but skips c. This is because NVDA has no way of knowing for sure that the "bc" button will be navigated as a single character. On the other hand, if you move by word, this works because "bc" is a word.

The only way to solve this would be to add some guesswork rules to the editing code; e.g. always treat an entire button as a character and never try to navigate the text inside it. There's a risk that this could be wrong in some cases, though; e.g. if some editor allowed you to edit the content of a button. Maybe the rule could be that we never try to navigate the text of anything with an author specified name (e.g. aria-label), assuming that the author overrode the content for a good reason in this case.

**Comment by Adriani90** [comment]
I navigated with left and right arrow in the to:, and cc: edit field, and then I explored the button name with review cursor and object nav. That‘s how I came accross the zero width space character and the button as separate childs.

I think your proposal is reasonable, as long as the user is still able to explore the button label character by character with the review cursor.

**Comment by jcsteh** [comment]
> I think your proposal is reasonable, as long as the user is still able to explore the button label character by character with the review cursor.

I think it's going to be one or the other, I'm afraid. As far as text is concerned, it's either a single compound character or multiple different characters. I don't think NVDA currently has a way to differentiate a character for caret vs review, although maybe some way could be added (e.g. another unit type).

**Comment by michaelDCurran** [comment]
It looks like the embedded recipient buttons are contenteditable="false" which is how they make them take up only one character position.
In theory, if NvDA sees the caret land on/inside a button or graphic with contenteditable="false" inside a contenteditable, it should probably read the button or graphic similar to focus (I.e. its name etc) rather than the first text character within.
@jcsteh  Do you have any thoughts on how NVDA could detect this? I don't think that Firefox or Chrome expose enough info via IAccessible2 to specifically identify contenteditable="false" within a contenteditable.

**Comment by jcsteh** [comment]
If something is contentEditable, even if it is a descendant, it should have the editable state (except for text leaf nodes). For example, the span below has the editable state:

`data:text/html,<div role="textbox" contenteditable>a<span role="button">b c</span>d`

In contrast, if a descendant overrides contentEditable to false, it should not get the editable state. For example, the span in this case doesn't have the editable state, even though its parent does:

`data:text/html,<div role="textbox" contenteditable>a<span contenteditable="false" role="button">b c</span>d`

This is true in both Firefox and Chrome.

**Comment by michaelDCurran** [comment]
ah yes, cool. That should be enough then.

**Comment by seanbudd** [comment]
Closed by #16870
